src/Control/Effect/Choose.hs

```diff
@@ -0,0 +1,95 @@
+{-# LANGUAGE DeriveGeneric, DeriveTraversable, FlexibleContexts, FlexibleInstances, MultiParamTypeClasses, RankNTypes, TypeOperators, UndecidableInstances #-}
+module Control.Effect.Choose
+( -- * Choose effect
+  Choose(..)
+, choose
+  -- * Choose carrier
+, runChoose
+, ChooseC(..)
+) where
+
+import Control.Applicative ((<|>), liftA2)
+import Control.Effect.Carrier
+import Control.Monad (join)
+import qualified Control.Monad.Fail as Fail
+import Control.Monad.Fix
+import Control.Monad.IO.Class
+import Control.Monad.Trans.Class
+import Data.Bool (bool)
+import Data.Maybe (fromJust)
+import GHC.Generics (Generic1)
+
+data Choose m k
+  = Choose (Bool -> m k)
+  deriving (Functor, Generic1)
+
+instance HFunctor Choose
+instance Effect   Choose
+
+-- | Nondeterministically choose between two computations.
+choose :: (Carrier sig m, Member Choose sig) => m a -> m a -> m a
+choose a b = send (Choose (bool b a))
+
+
+runChoose :: (m b -> m b -> m b) -> (a -> m b) -> ChooseC m a -> m b
+runChoose fork leaf m = runChooseC m fork leaf
+
+-- | A carrier for 'Choose' effects based on Ralf Hinze’s design described in [Deriving Backtracking Monad Transformers](https://www.cs.ox.ac.uk/ralf.hinze/publications/#P12).
+newtype ChooseC m a = ChooseC
+  { -- | A higher-order function receiving two continuations, respectively implementing choice and 'pure'.
+    runChooseC :: forall b . (m b -> m b -> m b) -> (a -> m b) -> m b
+  }
+  deriving (Functor)
+
+instance Applicative (ChooseC m) where
+  pure a = ChooseC (\ _ leaf -> leaf a)
+  {-# INLINE pure #-}
+  ChooseC f <*> ChooseC a = ChooseC $ \ fork leaf ->
+    f fork (\ f' -> a fork (leaf . f'))
+  {-# INLINE (<*>) #-}
+
+instance Monad (ChooseC m) where
+  ChooseC a >>= f = ChooseC $ \ fork leaf ->
+    a fork (runChoose fork leaf . f)
+  {-# INLINE (>>=) #-}
+
+instance Fail.MonadFail m => Fail.MonadFail (ChooseC m) where
+  fail s = lift (Fail.fail s)
+  {-# INLINE fail #-}
+
+instance MonadFix m => MonadFix (ChooseC m) where
+  mfix f = ChooseC $ \ fork leaf ->
+    mfix (runChoose (liftA2 Fork) (pure . Leaf)
+      . f . fromJust . fold (<|>) Just)
+    >>= fold fork leaf
```

Comment:
Like #197, this too suffers #209. I intend to fix this in a future PR (i.e. once I have any idea how to).